pkg/tfvars/openstack/rhcos_image.go

```diff
@@ -0,0 +1,139 @@
+package openstack
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/imageservice/v2/imagedata"
+	"github.com/gophercloud/gophercloud/openstack/imageservice/v2/imageimport"
+	"github.com/gophercloud/gophercloud/openstack/imageservice/v2/images"
+	"github.com/gophercloud/utils/openstack/clientconfig"
+	"github.com/sirupsen/logrus"
+)
+
+// uploadBaseImage creates a new image in Glance and uploads the RHCOS image there
+func uploadBaseImage(cloud string, localFilePath string, imageName string, clusterID string) error {
+	logrus.Debugln("Creating a Glance image for RHCOS...")
+
+	f, err := os.Open(localFilePath)
+	if err != nil {
+		return err
+	}
+
+	opts := clientconfig.ClientOpts{
+		Cloud: cloud,
+	}
+
+	conn, err := clientconfig.NewServiceClient("image", &opts)
+	if err != nil {
+		return err
+	}
+
+	imageCreateOpts := images.CreateOpts{
+		Name:            imageName,
+		ContainerFormat: "bare",
+		DiskFormat:      "raw",
+		Tags:            []string{fmt.Sprintf("openshiftClusterID=%s", clusterID)},
+		// TODO(mfedosin): add Description when gophercloud supports it.
+	}
+
+	img, err := images.Create(conn, imageCreateOpts).Extract()
+	if err != nil {
+		return err
+	}
+	logrus.Debugf("Image %s was created.", img.Name)
+
+	useImageImport, err := isImageImportSupported(cloud)
+	if err != nil {
+		return err
```

Comment:
Error checking whether glance supports image import should result is us assuming it does not rather than failing the image upload.

Reply:
I check what type of error Glance returns. If this is 404 because of the lack of the Discovery API, then `isImageImportSupported` ignores the error and returns `false, nil`. The function returns only real errors, for example, when Glance service is not available. In this case it's correct to stop the execution.

Reply:
The error could be anything, a 5xx because glance discovery API is broken, networking issue, ...
My point being, a failure in `isImageImportSupported()` is recoverable and should not stop execution: we can assume the cloud does not support glance direct image import.

Failure of glance image discovery at time T doesn't imply failure of glance image upload at time T+1. We're catching error of image upload, aren't we?

Reply:
I don't really like this decision. It leads to unpredictable system behavior when, for example, due to a short network failure, the installer believes that the image import is unavailable, but it is not.
As for me, so if Glance returns an error, we should inform the user about it, and do not hide it.

Reply:
There might be a higher-level concern on resilience, but I am unaware of that; and by default I am all for explicit errors.

If I'm not mistaken, the rationale for using image-import is to enable the upload of images in cases where the legacy upload would fail.

Thinking of this user story (legacy method fails), the user would probably be surprised of seeing that the upload fails despite his cluster having image-upload capability.

I'd really fail here with an explicit error, and let them try again from scratch.

---
As an alternative, I'd consider removing the check and _always_ uploading with image-import, and noisily falling back to the legacy upload in case of failure.